falcon/response.py

```diff
@@ -646,21 +646,23 @@ def set_headers(self, headers):
             :meth:`~.append_header` or :meth:`~.set_cookie`.
 
         Args:
-            headers (dict or list): A dictionary of header names and values
-                to set, or a ``list`` of (*name*, *value*) tuples. Both
-                *name* and *value* must be of type ``str`` and
+            headers (Iterable[[str, str]]): An iterable of ``[name, value]`` two-member
+                iterables, or a dict-like object that implements an ``items()`` method.
+                Both *name* and *value* must be of type ``str`` and
                 contain only US-ASCII characters.
 
                 Note:
-                    Falcon can process a list of tuples slightly faster
+                    Falcon can process an iterable of tuples slightly faster
                     than a dict.
 
         Raises:
-            ValueError: `headers` was not a ``dict`` or ``list`` of ``tuple``.
-
+            ValueError: `headers` was not a ``dict`` or ``list`` of ``tuple``
+                         or ``Iterable[[str, str]]``.
         """
 
-        if isinstance(headers, dict):
+        header_items = getattr(headers, 'items', None)
+
+        if header_items and callable(header_items):
```

Comment:
A less obvious further optimization for the mapping case, could we potentially just check `if callable(header_items):`?
```python
>>> callable(None)
False
```
This approach would perform slightly worse when just an iterable is provided though.
We would also need to add a `NOTE` explaining the above reasoning if we go this route.

OTOH maybe just checking for callable is also more correct? Consider the following example:
```python
>>> class HeaderFactory:
...     def __call__(self):
...         return (('X-one', '1'), ('X-two', '2'))
...     def __bool__(self):
...         return False
... 
>>> class Example:
...     def __init__(self):
...         self.items = HeaderFactory()
... 
>>> example = Example()
>>> example.items()
(('X-one', '1'), ('X-two', '2'))
>>> if example.items and callable(example.items):
...     raise NotImplementedError("I'm unreachable.")
... 
>>> callable(example.items)
True
```

If we keep the current solution we should IMHO document that `items()` is not called unless it is a truthy value.

Reply:
I think the approach with `items()` is a bit cleaner, and won't confused people, also the approach is straightforward and easy to understand, I prefer things that are easy for understanding.

Also, if we will switch to the suggested approach, I'm not sure that we'll take a lot of advantages.

Also, we will do the same thing with `getattr()`, since we're supporting `list of tuples` it also will require to check if the object has the attribute `items()` and etc.

So, I vote for the current approach;)

Reply:
@mikeyusko I think you might have misunderstood me. Probably my fault too, I got a bit carried away by those edge cases :thinking:

I was trying to compare the two following alternatives:
(current)
```python
if header_items and callable(header_items):
    # get items()
```
(proposed for discussion)
```python
if callable(header_items):
    # get items()
```

Reply:
@vytas7 Sorry for the misunderstanding:)

Yes, I also agree here, we can easy check `callable` and will be enough for our case, thanks!

I'll push the proposed changes tomorrow.

Reply:
@kgriffs any preference on `if header_items and callable(header_items):` vs `if callable(header_items):`?

Reply:
I'm fine with `if callable(header_items)`.